de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java

```diff
@@ -370,4 +379,25 @@ public boolean isWorkpackageProcessorEnabled(final Class<? extends IWorkpackageP
 				.stream()
 				.anyMatch(packageProcessor -> packageProcessorClassname.equals(packageProcessor.getClassname()));
 	}
+
+	@Override
+	public List<I_C_Queue_WorkPackage> retrieveUnprocessedWorkPackagesByEnqueuedRecord(
+			@NonNull final Class<? extends IWorkpackageProcessor> packageProcessorClass,
+			@NonNull final TableRecordReference recordRef)
+	{
+		final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+		final int workpackageProcessorId = retrievePackageProcessorIdByClass(packageProcessorClass);
+
+		return queryBL.createQueryBuilder(I_C_Queue_Element.class)
+				.addEqualsFilter(I_C_Queue_Element.COLUMNNAME_AD_Table_ID, recordRef.getAD_Table_ID())
+				.addEqualsFilter(I_C_Queue_Element.COLUMNNAME_Record_ID, recordRef.getRecord_ID())
+				.andCollect(I_C_Queue_Element.COLUMN_C_Queue_WorkPackage_ID)
+				.addNotEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_Processed, true)
+				.andCollect(I_C_Queue_WorkPackage.COLUMN_C_Queue_Block_ID)
+				.addEqualsFilter(I_C_Queue_Block.COLUMNNAME_C_Queue_PackageProcessor_ID, workpackageProcessorId)
```

Comment:
here you go from `C_Queue_WorkPackage` up to `C_Queue_Block` and then 
back down to `C_Queue_WorkPackage`.
Now you got *all* the WPs that have this C_Queue_Block_ID, even if not all of them have an element with the given `recordRef`.

We *often* have just one WP per block, but that's not the case in general.
So i think you need to modify this a bit.
IDK by heart to do this...maybe you can work with a subquery..ping me if you run into problems...

Reply:
done

Reply:
now the generated sql would look like 

```
SELECT ad_client_id,ad_issue_id,ad_org_id,ad_pinstance_id,ad_role_id,ad_user_id,ad_user_incharge_id,batchenqueuedcount,c_async_batch_id,c_queue_block_id,
( 
       SELECT c_queue_block.c_queue_packageprocessor_id 
       FROM   c_queue_block 
       WHERE  c_queue_block.c_queue_block_id=c_queue_workpackage.c_queue_block_id) AS c_queue_packageprocessor_id,c_queue_workpackage_id,created,createdby,errormsg,isactive,iserror,iserroracknowledged,isreadyforprocessing,lastdurationmillis,lastendtime, laststarttime,priority,processed,skipped_count,skipped_first_time,skipped_last_reason,skippedat,skiptimeoutmillis,updated,updatedby FROM c_queue_workpackage WHERE (
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_element 
           WHERE  ( 
                         c_queue_element.c_queue_workpackage_id = c_queue_workpackage.c_queue_workpackage_id
                  AND    c_queue_workpackage_id IS NOT NULL 
                  AND    (( 
                                       ad_table_id = 540270) 
                         AND    ( 
                                       record_id = 1000033)))) 
  ) 
  AND 
  ( 
    ( 
      processed <> 'Y' 
      OR 
      processed IS NULL 
    ) 
  ) 
  AND 
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_block 
           WHERE  ( 
                         c_queue_block.c_queue_block_id = c_queue_workpackage.c_queue_block_id 
                  AND    c_queue_block_id IS NOT NULL 
                  AND    (( 
                                       c_queue_packageprocessor_id = 540000)))) 
  ) 
)
```